src/util/opaque.h

```diff
@@ -0,0 +1,196 @@
+#pragma once
+
+#include <cstdlib>
+#include <memory>
+#include <optional>
+#include <utility>
+#include <variant>
+
+namespace bpftrace::util {
+
+// OpaqueValue corresponds to a blob of type-erased memory. It owns its own
+// underlying memory, taking a copy if necessary. This allows it to be used
+// to read maps, rings, and other locations which may not be permanent.
+//
+// This class does not really provide safety beyond first-order memory safety.
+class OpaqueValue {
+private:
+  struct OwnedBuffer {
+    OwnedBuffer(char *data) : data(data) {};
+    ~OwnedBuffer()
+    {
+      free(data);
+    }
+    char *data;
+  };
+  using SharedMemory = std::shared_ptr<OwnedBuffer>;
+  using Variant = std::variant<uintptr_t, SharedMemory>;
```

Comment:
Not really a fan of this name `Variant` just to have an alias for this small `std::variant` - bad for searches/groking